packages/valory/skills/abstract_round_abci/tests/test_common.py

```diff
@@ -51,3 +63,367 @@ def test_random_selection() -> None:
+                params=BaseParams(
+                    name="test",
+                    skill_context=MagicMock(),
+                    service_id="test_id",
+                    consensus=dict(max_participants=1),
+                    **irrelevant_config,
+                ),
+            ),
+        )
+
+    @pytest.mark.parametrize(
+        "return_value, expected_hash",
+        (
+            (MagicMock(performative=LedgerApiMessage.Performative.ERROR), None),
+            (MagicMock(state=MagicMock(body={"hash_key_is_not_in_body": ""})), None),
+            (
+                MagicMock(state=MagicMock(body={"hash": "test_randomness"})),
+                {
+                    "randomness": "d067b86fa5235e7e5225e8328e8faac5c279cbf57131d647e4da0a70df6d3d7b",
+                    "round": 0,
+                },
+            ),
+        ),
+    )
+    def test_failsafe_randomness(
+        self, return_value: MagicMock, expected_hash: Optional[str]
+    ) -> None:
+        """Test `failsafe_randomness`."""
+        gen = self.randomness_behaviour.failsafe_randomness()
+
+        with mock.patch.object(
+            DummyRandomnessBehaviour,
+            "get_ledger_api_response",
+            dummy_generator(return_value),
+        ):
+            next(gen)
+            try:
+                next(gen)
+            except StopIteration as e:
+                assert e.value == expected_hash
+            else:
+                AssertionError(
+                    "`get_ledger_api_response`'s generator should have been exhausted."
+                )
+
+    @pytest.mark.parametrize("randomness_response", ("test", None))
+    @pytest.mark.parametrize("verified", (True, False))
+    def test_get_randomness_from_api(
+        self, randomness_response: Optional[str], verified: bool
+    ) -> None:
+        """Test `get_randomness_from_api`."""
+        # create a dummy `process_response` for `MagicMock`ed `randomness_api`
+        self.randomness_behaviour.context.randomness_api.process_response = (
+            lambda res: res + "_processed" if res is not None else None
+        )
+        gen = self.randomness_behaviour.get_randomness_from_api()
+
+        with mock.patch.object(
+            DummyRandomnessBehaviour,
+            "get_http_response",
+            dummy_generator(randomness_response),
+        ), mock.patch.object(
+            VerifyDrand,
+            "verify",
+            return_value=(verified, "Error message."),
+        ):
+            next(gen)
+            try:
+                next(gen)
+            except StopIteration as e:
+                if randomness_response is None or not verified:
+                    assert e.value is None
+                else:
+                    assert e.value == randomness_response + "_processed"
+            else:
+                AssertionError(
+                    "`get_randomness_from_api`'s generator should have been exhausted."
+                )
```

Comment:
`raise` is missing here

Reply:
Added in ac8d829ef855aed9422b7493ccd133ed7c85dd66.